keps/sig-node/2400-node-swap/README.md

```diff
@@ -0,0 +1,836 @@
+virtio-fs, …). Swap helps to not request much more memory to deal with short
+term worst-case scenarios.
+
+With virtualization, clusters are typically provisioned based on the workloads’
+memory consumption, and any infrastructure container overhead is overcommitted.
+This overhead could be safely swapped out.
+
+- Required for live migration of VMs
+  https://github.com/kubernetes/kubernetes/issues/53533#issuecomment-754878431
+
+This user story is addressed by scenario 2, and could benefit from 3.
+
+### Notes/Constraints/Caveats (Optional)
+
+In changing the CRI, we must ensure that container runtime downstreams are able
+to support the new configurations.
+
+We considered adding parameters for both per-workload `memory-swap` and
+`swappiness`. These are documented as part of the Open Containers [runtime
+specification] for Linux memory configuration. Since `memory-swap` is a
+per-workload parameter, and `swappiness` is optional and can be set globally,
+we are choosing to only expose `memory-swap` which will adjust swap available
+to workloads.
+
+Since we are not currently setting `memory-swap` in the CRI, the current
+default behaviour when `--fail-swap-on=false` is set is to allocate the same
+amount of swap for a workload as memory requested. We will update the default
+to not permit the use of swap by setting `memory-swap` equal to `limit`.
+
+[runtime specification]: https://github.com/opencontainers/runtime-spec/blob/1c3f411f041711bbeecf35ff7e93461ea6789220/config-linux.md#memory
+
+### Risks and Mitigations
+
+Having swap available on a system reduces predictability. Swap's performance is
+worse than regular memory, sometimes by many orders of magnitude, which can
+cause unexpected performance regressions. Furthermore, swap changes a system's
+behaviour under memory pressure, and applications cannot directly control what
+portions of their memory usage are swapped out. Since enabling swap permits
+greater memory usage for workloads in Kubernetes that cannot be predictably
+accounted for, it also increases the risk of noisy neighbours and unexpected
+packing configurations, as the scheduler cannot account for swap memory usage.
+
+This risk is mitigated by preventing any workloads from using swap by default,
+even if swap is enabled and available on a system. This will allow a cluster
+administrator to test swap utilization just at the system level without
+introducing unpredictability to workload resource utilization.
+
+Additionally, we will mitigate this risk by determining a set of metrics to
+quantify system stability and then gathering test and production data to
+determine if system stability changes when swap is available to the system
+and/or workloads in a number of different scenarios.
+
+Since swap provisioning is out of scope of this proposal, this enhancement
+poses low risk to Kubernetes clusters that will not enable swap.
+
+## Design Details
+
+We summarize the implementation plan as following:
```

Comment:
just to clarify, this is saying that all resources below kubepods cgroup have access to same amount of scratch swap meaning swap itself is an overcommitted resource?

Reply:
Yes.